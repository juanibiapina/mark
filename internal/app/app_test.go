package app

import (
	"fmt"
	"os"
	"path"
	"testing"

	tea "github.com/charmbracelet/bubbletea/v2"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func makeApp(t *testing.T, cwd string) App {
	app, err := MakeApp(cwd)
	assert.Nil(t, err)
	return app
}

func bareApp(t *testing.T) App {
	cwd := t.TempDir()
	app := makeApp(t, cwd)
	model, _ := app.Update(tea.WindowSizeMsg{Width: 64, Height: 16})
	return model.(App)
}

// handleMessage processes messages like in a tea.Program.
// This is needed for handling internal messages like tea.BatchMsg.
func handleMessage(t *testing.T, model tea.Model, msg tea.Msg) tea.Model {
	switch msg := msg.(type) {
	case tea.QuitMsg:
		t.Fatal("unexpected quit message")
	case tea.BatchMsg:
		for _, cmd := range msg {
			m := cmd()
			model, _ = model.Update(m)
		}
	default:
		model, _ = model.Update(msg)
	}
	return model
}

func update(app App, msg tea.Msg) App {
	model, _ := app.Update(msg)
	return model.(App)
}

// render a tea.Model to a string assuming it implements tea.ViewModel.
func render(t *testing.T, model tea.Model) string {
	vmodel, ok := model.(tea.ViewModel)
	require.True(t, ok)

	v := vmodel.View()
	return v
}

func key(code rune) tea.Msg {
	return tea.KeyPressMsg{Code: code, Text: string(code)}
}

func TestMain(m *testing.M) {
	v := m.Run()

	snaps.Clean(m, snaps.CleanOpts{Sort: true})

	os.Exit(v)
}

func TestApp(t *testing.T) {
	t.Parallel()

	t.Run("Initialization", func(t *testing.T) {
		cwd := path.Join("testdata", "cwd")
		app := makeApp(t, cwd)

		v := app.View()
		assert.Equal(t, "Initializing...", v)

		cmd := app.Init()
		assert.NotNil(t, cmd)

		model, _ := app.Update(tea.WindowSizeMsg{Width: 64, Height: 16})

		msg := cmd()
		model = handleMessage(t, model, msg)

		v = render(t, model)

		snaps.MatchStandaloneSnapshot(t, v)
	})

	t.Run("Error", func(t *testing.T) {
		app := bareApp(t)

		err := fmt.Errorf("test error")

		model, cmd := app.Update(errMsg{err: err})

		assert.Equal(t, tea.QuitMsg{}, cmd())
		app = model.(App)
		require.ErrorIs(t, app.Err(), err)
	})

	t.Run("messages", func(t *testing.T) {
		t.Run("replyMessage", func(t *testing.T) {
			app := bareApp(t)

			model, cmd := app.Update(replyMessage("test"))
			assert.NotNil(t, cmd)

			v := render(t, model)
			snaps.MatchStandaloneSnapshot(t, v)
		})
	})

	t.Run("focus", func(t *testing.T) {
		app := bareApp(t)
		focuses := []Focused{FocusedInput, FocusedThreadList, FocusedThread, FocusedInput}

		for _, expectedFocus := range focuses {
			require.Equal(t, expectedFocus, app.focused)
			v := app.View()
			snaps.MatchSnapshot(t, v)
			app = update(app, key(tea.KeyTab))
		}
	})

	t.Run("input", func(t *testing.T) {
		app := bareApp(t)

		app = update(app, key('h'))
		app = update(app, key('e'))
		app = update(app, key('y'))

		v := app.View()
		snaps.MatchSnapshot(t, v)

		app = update(app, key(tea.KeyEnter))
		v = app.View()
		snaps.MatchSnapshot(t, v)
	})
}
