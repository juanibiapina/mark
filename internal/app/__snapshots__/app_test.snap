
[TestApp/focus - 1]
â•­â”€Contextâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®â•­â”€Messagesâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚[38;2;98;98;98mNo Context.[m        â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
[32mâ•­â”€[0m[1;32mMessage Assistant[m[32mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[m
[32mâ”‚[m[37m[m[7;37m [m                                                             [32mâ”‚[m
[32mâ”‚[m[30m                                                              [m[32mâ”‚[m
[32mâ”‚[m[30m                                                              [m[32mâ”‚[m
[32mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[m
---

[TestApp/focus - 2]
[32mâ•­â”€[0m[1;32mContext[m[32mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[mâ•­â”€Messagesâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
[32mâ”‚[m[38;2;98;98;98mNo Context.[m        [32mâ”‚[mâ”‚                                         â”‚
[32mâ”‚[m                   [32mâ”‚[mâ”‚                                         â”‚
[32mâ”‚[m                   [32mâ”‚[mâ”‚                                         â”‚
[32mâ”‚[m                   [32mâ”‚[mâ”‚                                         â”‚
[32mâ”‚[m                   [32mâ”‚[mâ”‚                                         â”‚
[32mâ”‚[m                   [32mâ”‚[mâ”‚                                         â”‚
[32mâ”‚[m                   [32mâ”‚[mâ”‚                                         â”‚
[32mâ”‚[m                   [32mâ”‚[mâ”‚                                         â”‚
[32mâ”‚[m                   [32mâ”‚[mâ”‚                                         â”‚
[32mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
â•­â”€Message Assistantâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚[37m[m[7;37m [m                                                             â”‚
â”‚[30m                                                              [mâ”‚
â”‚[30m                                                              [mâ”‚
â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
---

[TestApp/focus - 3]
â•­â”€Contextâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[32mâ•­â”€[0m[1;32mMessages[m[32mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[m
â”‚[38;2;98;98;98mNo Context.[m        â”‚[32mâ”‚[m                                         [32mâ”‚[m
â”‚                   â”‚[32mâ”‚[m                                         [32mâ”‚[m
â”‚                   â”‚[32mâ”‚[m                                         [32mâ”‚[m
â”‚                   â”‚[32mâ”‚[m                                         [32mâ”‚[m
â”‚                   â”‚[32mâ”‚[m                                         [32mâ”‚[m
â”‚                   â”‚[32mâ”‚[m                                         [32mâ”‚[m
â”‚                   â”‚[32mâ”‚[m                                         [32mâ”‚[m
â”‚                   â”‚[32mâ”‚[m                                         [32mâ”‚[m
â”‚                   â”‚[32mâ”‚[m                                         [32mâ”‚[m
â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[32mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[m
â•­â”€Message Assistantâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚[37m[m[7;37m [m                                                             â”‚
â”‚[30m                                                              [mâ”‚
â”‚[30m                                                              [mâ”‚
â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
---

[TestApp/focus - 4]
â•­â”€Contextâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®â•­â”€Messagesâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚[38;2;98;98;98mNo Context.[m        â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
[32mâ•­â”€[0m[1;32mMessage Assistant[m[32mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[m
[32mâ”‚[m[37m[m[7;37m [m                                                             [32mâ”‚[m
[32mâ”‚[m[30m                                                              [m[32mâ”‚[m
[32mâ”‚[m[30m                                                              [m[32mâ”‚[m
[32mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[m
---

[TestApp/input - 1]
â•­â”€Contextâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®â•­â”€Messagesâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚[38;2;98;98;98mNo Context.[m        â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
[32mâ•­â”€[0m[1;32mMessage Assistant[m[32mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[m
[32mâ”‚[m[37m[mhey[7;37m [m                                                          [32mâ”‚[m
[32mâ”‚[m[30m                                                              [m[32mâ”‚[m
[32mâ”‚[m[30m                                                              [m[32mâ”‚[m
[32mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[m
---

[TestApp/input - 2]
â•­â”€Contextâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®â•­â”€Messagesâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚[38;2;98;98;98mNo Context.[m        â”‚â”‚                                      heyâ”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â”‚                   â”‚â”‚                                         â”‚
â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
[32mâ•­â”€[0m[1;32mMessage Assistant[m[32mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[m
[32mâ”‚[m[37m[m[7;37m [m                                                             [32mâ”‚[m
[32mâ”‚[m[30m                                                              [m[32mâ”‚[m
[32mâ”‚[m[30m                                                              [m[32mâ”‚[m
[32mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[m
---
